src/igraph/constructors/from_object.py

```diff
@@ -0,0 +1,462 @@
+      with the extra items in the edge list (starting from index 2, since
+      the first two items are the source and target vertices). C{None}
+      means that only the source and target vertices will be extracted
+      from each item. If you pass a string here, it will be wrapped in
+      a list for convenience.
+    @param weights: alternative way to specify that the graph is
+      weighted. If you set C{weights} to C{true} and C{edge_attrs} is
+      not given, it will be assumed that C{edge_attrs} is C{["weight"]}
+      and igraph will parse the third element from each item into an
+      edge weight. If you set C{weights} to a string, it will be assumed
+      that C{edge_attrs} contains that string only, and igraph will
+      store the edge weights in that attribute.
+    @return: the graph that was constructed
+    """
+    if edge_attrs is None:
+        if not weights:
+            edge_attrs = ()
+        else:
+            if not isinstance(weights, str):
+                weights = "weight"
+            edge_attrs = [weights]
+    else:
+        if weights:
+            raise ValueError(
+                "`weights` must be False if `edge_attrs` is " "not None"
+            )
+
+    if isinstance(edge_attrs, str):
+        edge_attrs = [edge_attrs]
+
+    # Set up a vertex ID generator
+    idgen = UniqueIdGenerator()
+
+    # Construct the edges and the edge attributes
+    edge_list = []
+    edge_attributes = {}
+    for name in edge_attrs:
+        edge_attributes[name] = []
+
+    for item in edges:
+        edge_list.append((idgen[item[0]], idgen[item[1]]))
+        for index, name in enumerate(edge_attrs, 2):
+            try:
+                edge_attributes[name].append(item[index])
+            except IndexError:
+                edge_attributes[name].append(None)
+
+    # Set up the "name" vertex attribute
+    vertex_attributes = {}
+    vertex_attributes[vertex_name_attr] = list(idgen.values())
+    n = len(idgen)
+
+    # Construct the graph
+    return cls(n, edge_list, directed, {}, vertex_attributes, edge_attributes)
+
+
+def construct_graph_from_sequence_dict(
+    cls,
+    edges,
+    directed=False,
+    vertex_name_attr="name",
+):
+    """Constructs a graph from a dict-of-sequences representation.
+
+    This function is used to construct a graph from a dictionary of
+    sequences (e.g. of lists). For each key x, its corresponding value is
+    a sequence of multiple objects: for each y, the edge (x,y) will be
+    created in the graph. x and y must be either one of:
+
+    - two integers: the vertices with those ids will be connected
+    - two strings: the vertices with those names will be connected
+
+    If names are used, the order of vertices is not guaranteed, and each
+    vertex will be given the vertex_name_attr attribute.
+    """
+    # Check if we are using names or integers
+    for item in edges:
+        break
+    if not isinstance(item, (int, str)):
```

Comment:
This will blow up if `edges` is an empty iterable; it should be handled separately.

I have found the following helper function useful in such cases:

```
def first(iterable, default=None):
    for item in iterable:
        return item
    return default
```

and then you can simply write:

```
if not isinstance(first(edges, default=0), (int, str)):
    raise ValueError("Keys must be integers or strings")
```

and this will handle the case of an empty iterable correctly.

Reply:
Good catch, I fixed the corner case now